lib/finch/http1/conn.ex

```diff
@@ -174,33 +175,83 @@ defmodule Finch.Conn do
         Telemetry.stop(:recv, start_time, metadata, extra_measurements)
         {:ok, %{conn | mint: mint}, acc}
 
+      {:error, _mint, error = :response_body_too_large} ->
+        metadata = Map.put(metadata, :error, error)
+        Telemetry.stop(:recv, start_time, metadata, extra_measurements)
+        {:error, close(conn), Error.exception(error)}
+
       {:error, mint, error} ->
         metadata = Map.put(metadata, :error, error)
         Telemetry.stop(:recv, start_time, metadata, extra_measurements)
         {:error, %{conn | mint: mint}, error}
     end
   end
 
-  defp receive_response([], acc, fun, mint, ref, timeout) do
+  defp receive_response(entries, acc, fun, mint, ref, timeout, max_body, body_sum \\ 0)
+
+  defp receive_response([], acc, fun, mint, ref, timeout, max_body, body_sum) do
     case HTTP1.recv(mint, 0, timeout) do
       {:ok, mint, entries} ->
-        receive_response(entries, acc, fun, mint, ref, timeout)
+        receive_response(entries, acc, fun, mint, ref, timeout, max_body, body_sum)
 
       {:error, mint, error, _responses} ->
         {:error, mint, error}
     end
   end
 
-  defp receive_response([entry | entries], acc, fun, mint, ref, timeout) do
+  defp receive_response([entry | entries], acc, fun, mint, ref, timeout, max_body, body_sum) do
     case entry do
       {:status, ^ref, value} ->
-        receive_response(entries, fun.({:status, value}, acc), fun, mint, ref, timeout)
+        receive_response(
+          entries,
+          fun.({:status, value}, acc),
+          fun,
+          mint,
+          ref,
+          timeout,
+          max_body,
+          body_sum
+        )
 
       {:headers, ^ref, value} ->
-        receive_response(entries, fun.({:headers, value}, acc), fun, mint, ref, timeout)
+        with true <- is_integer(max_body),
+             {_header_name, content_length_str} <- List.keyfind(value, "content-length", 0),
+             {content_length, ""} = Integer.parse(content_length_str),
+             true <- content_length > max_body do
+          {:error, mint, :response_body_too_large}
+        else
+          _ ->
+            receive_response(
+              entries,
+              fun.({:headers, value}, acc),
+              fun,
+              mint,
+              ref,
+              timeout,
+              max_body,
+              body_sum
+            )
+        end
 
       {:data, ^ref, value} ->
-        receive_response(entries, fun.({:data, value}, acc), fun, mint, ref, timeout)
+        body_sum = :erlang.iolist_size(value) + body_sum
```

Comment:
I think this is debatable because the cost is tiny compared to the HTTP request itself, and the branch operation is likely to be more costly than the calculation ([pipeline stall](https://en.wikipedia.org/wiki/Pipeline_stall)). Also, I found it helpful when using `:recon_trace.calls` to track how much data was transferred by looking at the body_sum parameter.